### PR TITLE
feat(express): add graceful exception handling

### PR DESCRIPTION
**Pull request by dmbch**
This PR introduces a new config option: `gracePeriod`. In case of uncaught asynchronous middleware errors or upon receival of a `SIGTERM` signal, the server's `close` method will now be called before exiting the process. If the server fails to finish answering pending requests before `gracePeriod` is over, it will be terminated.

### CODE REVIEW

File: packages/express/lib/run.js [hunk 1]
```diff
@@ -22,6 +22,23 @@ const createServer = (app, https) =>
...
     : createHTTPServer(app);
+  return app.locals.server;
+};
+
+const shutdown = (app) => {
+  app.locals.shuttingDown = true;
+  return Promise.race([
+    new Promise((resolve, reject) =>
+      setTimeout(
+        () => reject(new Error('timeout')),
```
Comment (dmbch) [inline packages/express/lib/run.js]: We ought to move `gracePeriod` to config, I think.

File: packages/express/lib/run.js [hunk 2]
```diff
@@ -8,8 +8,8 @@ const { createServer: createHTTPSServer } = require('https');
 
 const portfinder = require('portfinder');
 
-const createServer = (app, https) =>
-  https
+const createServer = (app, https) => {
+  app.locals.server = https
```
Comment (dmbch) [inline packages/express/lib/run.js]: Can we get away with not exposing `server` in `app.locals`?

File: packages/express/lib/run.js [hunk 3]
```diff
@@ -30,16 +47,25 @@ const getPort = (port) => {
...
-    server.listen(port, host, (error) => {
-      if (error) {
-        server.close();
-        handleError(error);
-      } else {
-        inspectServer(server);
-      }
-    })
+  domain.on('error', (error) =>
+    shutdown(app).then(() => handleError(error), () => handleError(error))
```
Comment (herschel666) [inline packages/express/lib/run.js]: If shutting down the app on `SIGTERM` errors, will this trigger the `'error'`-callback of the `domain` and thus invoke a second shutdown of the app?

Comment (herschel666) [reply packages/express/lib/run.js]: Forget what I wrote. I got `handleError()` wrong.

File: packages/express/README.md [hunk 4]
```diff
@@ -82,9 +84,9 @@ module.exports = class MyMixin extends Mixin {
...
 
-### `inspectServer(app, target)` ([sequence](https://github.com/untool/mixinable/blob/master/README.md#defineparallel))
+### `inspectServer(server)` ([sequence](https://github.com/untool/mixinable/blob/master/README.md#defineparallel))
 
-This hook will give you a running, i.e. listening, instance of [`http.Server`](https://nodejs.org/api/http.html#http_class_http_server) or [`https.Server`](https://nodejs.org/api/https.html#https_class_https_server), depending on your `https` setting. The second argument, `target`, will only ever be one of `develop` and `serve`. You can, for example, use this hook to register your server with an external load balancing system.
+This hook will give you a running, i.e. listening, instance of [`http.Server`](https://nodejs.org/api/http.html#http_class_http_server) or [`https.Server`](https://nodejs.org/api/https.html#https_class_https_server), depending on your `https` setting. This server will emit an additional `shutdown` event in case a gracefull shutdown is triggered.
```
[the remark this answers is not in the document]
Comment (dmbch) [reply packages/express/README.md]: Done, thanks.

File: packages/express/README.md [hunk 5]
```diff
@@ -157,3 +160,21 @@ This is the file system path, i.e. subfolder, your application's assets will be
...
+
+The amount of time (in milliseconds) to wait after receiving a [`SIGTERM`](https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html) signal or catching an unhandled middleware exception and before killing the server completely.
+
+```json
+{
+  "gracePeriod": 60000
+}
+```
+
+server completely.
```
[the remark this answers is not in the document]
Comment (dmbch) [reply packages/express/README.md]: Resolved, thanks.

**Pull request merged**